tests/lib/ocpp/v16/test_smart_charging_handler.cpp

```diff
@@ -0,0 +1,764 @@
+                               {}};
+    }
+
+    SmartChargingHandler* createSmartChargingHandler() {
+        // connectors[0] = std::make_shared<Connector>(Connector{0});
+        // connectors[1] = std::make_shared<Connector>(Connector{1});
+
+        // const std::string chargepoint_id = "1";
+        // const fs::path database_path = "na";
+        // const fs::path init_script_path = "na";
+
+        // std::shared_ptr<DatabaseHandlerMock> database_handler =
+        //     std::make_shared<DatabaseHandlerMock>(chargepoint_id, database_path, init_script_path);
+
+        // auto handler = new SmartChargingHandler(connectors, database_handler, true);
+
+        return createSmartChargingHandler(0);
+    }
+
+    // 
+    SmartChargingHandler* createSmartChargingHandler(const int number_of_connectors) {
+        for (int i = 0; i <= number_of_connectors; i++) {
+            // connectors[i] = std::make_shared<Connector>(Connector{i});
+            addConnector(i);
+        }
+
+        const std::string chargepoint_id = "1";
+        const fs::path database_path = "na";
+        const fs::path init_script_path = "na";
+
+        std::shared_ptr<DatabaseHandlerMock> database_handler =
+            std::make_shared<DatabaseHandlerMock>(chargepoint_id, database_path, init_script_path);
+
+        auto handler = new SmartChargingHandler(connectors, database_handler, true);
+
+        return handler;
+    }
+
+
+    SmartChargingHandler* createSmartChargingHandlerWithChargePointMaxProfile() {
+        auto profile = createChargingProfile(createChargeSchedule(ChargingRateUnit::A));
+        const std::vector<ChargingRateUnit>& charging_schedule_allowed_charging_rate_units{ChargingRateUnit::A};
+        auto handler = createSmartChargingHandler(10);
+
+        profile.chargingProfilePurpose = ChargingProfilePurposeType::ChargePointMaxProfile;
+        profile.chargingProfileKind = ChargingProfileKindType::Absolute;
+        const int connector_id = 0;
+        const int max_charging_profiles_installed = 10;
+        bool is_profile_valid = handler->validate_profile(profile, connector_id, ignore_no_transaction, profile_max_stack_level,
+                                            max_charging_profiles_installed, charging_schedule_max_periods,
+                                            charging_schedule_allowed_charging_rate_units);
+        handler->add_charge_point_max_profile(profile);
+
+        return handler;
+    }
+
+    // Default values used within the tests
+    std::map<int32_t, std::shared_ptr<Connector>> connectors;
+    std::shared_ptr<DatabaseHandler> database_handler;
+
+    const int connector_id = 1;
+    bool ignore_no_transaction = true;
+    const int profile_max_stack_level = 10;
+    const int max_charging_profiles_installed = 20;
+    const int charging_schedule_max_periods = 10;
+};
+
+/**
+ * PB01 Valid Profile
+ *
+ * Happy path simple test
+ */
+TEST_F(ChargepointTestFixture, ValidateProfile) {
```

Comment:
Since these are unit tests, I understand that they are linked to code functions. But it might also be helpful to indicate which functional tests the functions map to (and the unit tests influence) (e.g. A05)?

Reply:
Since VSCode can find all that information with a right mouse click, I feel that adding these comments is unnecessary and can quickly become out of date if not careful.

Reply:
- not everybody uses VSCode
- I am not sure how VSCode would find the match to the OCPP specification which is a pdf document that is not in the code.

Maybe you can share some screenshots to clarify?

Reply:
@shankari The naming convention of the test is showing what functionality is being covered at a highlevel. I am not sure we have the test setup to the level of granularity of a functional requirement in the spec. The SUT itself is covering the functional requirement. Each conditional in there is probably more accurate for labeling like this. In general though comments like that can become stale pretty quickly.

In relation to the VS code context of finding the reference to a method call. That should apply to most modern IDEs.

Reply:
> In general though comments like that can become stale pretty quickly.

I agree that in general it is better to not duplicate information (see https://github.com/EVerest/everest-utils/pull/111#discussion_r1434250662) 

But given that the goal of this library is to implement the spec, and the functional spec is in a pdf, I think it would be helpful to have explicit comments that link code to (potentially multiple) cases in the specification.

I don't see that they will get stale quickly since the spec is fixed - if we refactor code, part of the refactor is to keep these up to date.

I am open to suggestions on where the mapping can be maintained, but I think that it shouid be maintained.